pkg/runner/test_runner.go

```diff
@@ -6,282 +6,778 @@
-	r.completedRouting = make(map[string]error)
-	r.completedTargets = make(map[*target.Target]error)
-	return &r
+// stepReader receives results from the step's output channel and forwards them to the appropriate target runners.
+func (tr *testRunner) stepReader(ctx statectx.Context, ss *stepState) {
+	ss.log.Debugf("%s: step reader active", ss)
+	var err error
+	outCh := ss.outCh
+loop:
+	for {
+		select {
+		case tgt, ok := <-outCh:
+			if !ok {
+				ss.log.Debugf("%s: out chan closed", ss)
+				// At this point we may still have an error to report,
+				// wait until error channel is emptied too.
+				outCh = make(chan *target.Target)
+				tr.safeCloseErrCh(ss)
+				continue loop
+			}
+			if err = tr.reportTargetResult(ctx, ss, tgt, nil); err != nil {
+				break loop
+			}
+		case res, ok := <-ss.errCh:
+			if !ok {
+				ss.log.Debugf("%s: err chan closed", ss)
+				break loop
+			}
+			if err = tr.reportTargetResult(ctx, ss, res.Target, res.Err); err != nil {
+				break loop
+			}
+		case <-ctx.Done():
+			ss.log.Debugf("%s: canceled 3, draining", ss)
+			for {
+				select {
+				case <-outCh:
+					break loop
+				case <-ss.errCh:
+					break loop
+				case <-time.After(tr.shutdownTimeout):
+					tr.mu.Lock()
+					if ss.runErr == nil {
+						ss.runErr = &cerrors.ErrTestStepsNeverReturned{}
+					}
+					tr.mu.Unlock()
+				}
+			}
+		}
+	}
+	tr.mu.Lock()
+	defer tr.mu.Unlock()
+	if ss.runErr == nil && err != nil {
+		ss.runErr = err
+	}
+	if ss.stepRunning && ss.runErr == nil {
+		// This means that plugin closed its channels before leaving.
+		ss.runErr = &cerrors.ErrTestStepClosedChannels{StepName: ss.sb.TestStepLabel}
+	}
+	ss.readerRunning = false
+	ss.log.Debugf("%s: step reader finished, %t %t %v", ss, ss.stepRunning, ss.readerRunning, ss.runErr)
+	tr.cond.Signal()
 }
 
-// CompletedTargets returns a map that associates each target with its returning error.
-// If the target succeeded, the error will be nil
-func (r *State) CompletedTargets() map[*target.Target]error {
-	return r.completedTargets
+// checkStepRunnersLocked checks if any step runner has encountered an error.
+func (tr *testRunner) checkStepRunners() error {
+	for _, ss := range tr.steps {
+		if ss.runErr != nil {
+			return ss.runErr
+		}
+	}
+	return nil
 }
 
-// CompletedRouting returns a map that associates each routing block with its returning error.
-// If the routing block succeeded, the error will be nil
-func (r *State) CompletedRouting() map[string]error {
-	return r.completedRouting
+// runMonitor monitors progress of targets through the pipeline
+// and closes input channels of the steps to indicate that no more are expected.
+// It also monitors steps for critical errors and cancels the whole run.
+// Note: input channels remain open when cancellation is requested,
+// plugins are expected to handle it explicitly.
+func (tr *testRunner) runMonitor() error {
+	tr.log.Debugf("monitor: active")
+	tr.mu.Lock()
+	defer tr.mu.Unlock()
```

Comment:
I see you lock the mutex for the whole runMonitor execution. How `targetRunners` works in this case?